### DISCUSSION TITLE
How do I use native APIs on mobile

### DISCUSSION THREAD

**Discussion by SaverinOnRails** [AvaloniaUI/Avalonia#11730] [Q&A] [answered]
This is probably an amateur question but im still uncertain on how to use native apis like status bar and battery. How exactly am I supposed to interact with them?

**Comment by maxkatz6** [comment]
Your mobile projects have "net6.0-ios" and "net6.0-android" targets, which gives you access to the native APIs.
For example, MS documentation for these
https://learn.microsoft.com/en-us/dotnet/api/android.app.activity?view=xamarin-android-sdk-13&viewFallbackFrom=net-6.0
https://learn.microsoft.com/en-us/dotnet/api/foundation.nsnotification?view=xamarin-ios-sdk-12

Or you can use Android/iOS docs directly (which is preferred as they usually have more remarks). Microsoft has mapped native APIs on .NET types.

**Reply by SaverinOnRails** [reply]
Hi. Thanks for the reply. I understand i can use the native android library from the Android project , but how can i communicate with it from the avalonia app. Can you explain how i can use the apis from say... a viewmodel class?

**Reply by timunie** [reply]
You can for example create an Interface or a Service, that the ViewModel can call. Register it on App start for each relevant platform.

**Reply by SaverinOnRails** [reply]
> You can for example create an Interface or a Service, that the ViewModel can call. Register it on App start for each relevant platform. 

But to do that wouldn't I need to add a reference to the android project, I cant add a reference since the android package already has a reference to the avalonia project

**Reply by timunie** [reply, marked as answer]
No, the other way around. You create the interfache in your main Avalonia Project

```
MyAvaloniaProject
   - Interfaces
        - IBatteryService
          * double GetBatteryCharge()
          * void SafeBattery()
          * ...
    - static RegisteredServices
       * static IBatteryService BatteryService {get; set;}
```

and in the `MyAvaloniaProject.Android` project you need to implement that interface

```cs
using MyAvaloniaProject

public class AndroidBatteryService : IBatteryService 
{
   ... 
}
```

now you can register that interface on Android startup

```cs
MyAndroidProject.RegisteredServices.BatteryService = new AndroidBatteryService();
```

I hope it gets a bit clearer now.

**Reply by SaverinOnRails** [reply]
Hi again and thanks for your reply. I think i'm beginning to understand however i am getting a nullReferenceException where i call the service. Could this be becuase i didnt register it at the right time? 
```
//null reference exception is thrown here
 public MainViewModel()
        {
            Debug.WriteLine(RegisteredServices.LocationService.CurrentLocation().Lat);
        }
```
```
//this is where i register the service. In MainActivity.cs within the android project.
 protected override void OnCreate(Bundle savedInstanceState)
        {
           AvalonWeather.RegisteredServices.LocationService = new AndroidLocation();
            base.OnCreate(savedInstanceState); 
        }
```
```
public static class RegisteredServices
{
    public static ILocation LocationService { get; set; }
}
```
-----EDIT------
Got it to work by initialising in the splash activity:
```
  protected override AppBuilder CustomizeAppBuilder(AppBuilder builder)
        {
            RegisteredServices.LocationService = new AndroidLocation();
            return base.CustomizeAppBuilder(builder)
                .UseReactiveUI();
        }
```
Thanks for all your help  @timunie  and @maxkatz6
